core-backend/src/funcs.rs

```diff
@@ -128,14 +87,20 @@ impl From<u32> for SyscallReturnValue {
     }
 }
 
-pub(crate) trait SyscallContext: Sized {
+pub(crate) trait SyscallContext: Sized + Copy {
     fn from_args(args: &[Value]) -> Result<(Self, &[Value]), HostError>;
+
+    fn gas(&self) -> Option<Gas>;
 }
 
 impl SyscallContext for () {
     fn from_args(args: &[Value]) -> Result<(Self, &[Value]), HostError> {
         Ok(((), args))
     }
+
+    fn gas(&self) -> Option<Gas> {
```

Comment:
I actually do not see the reason of such changes (for `run_fallible` and `run_any`, too). Could you clarify, please?

Reply:
Getting gas from SyscallContext is necessary to set the amount of gas in Ext::decrease_current_counter_to() before calling Builder::build() and Syscall::execute(), because MemoryAccessRegistry::pre_process() consumes gas that we would not otherwise include

Reply:
All in all I don't know what would be the best way to do it, let it be your variant =)